src/app/beer_garden/api/http/authentication/login_handlers/ldap.py

```diff
@@ -0,0 +1,122 @@
+
+        Returns:
+            bool: True if the password matches, False otherwise
+        """
+        if conn.result["description"] == "success":
+            return True
+        return False
+
+    def get_user_roles(self, conn: Connection, username: str):
+        """Checks the users roles against the provided"""
+        groups = set(config.get("ldap.default_user_roles"))
+        roles = []
+        conn.search(
+            config.get("ldap.roles_search_base"),
+            f"(&(objectclass=groupOfNames)(member={self.get_user_dn(username)}))",
+            attributes=["cn"],
+        )
+        for entry in conn.entries:
+            groups.add(entry["cn"].value)
+
+        for group in groups:
+            try:
+                roles.append(get_role(role_name=group))
+            except DoesNotExist:
+                pass
+
+        logger.info(
+            f"Updating {username} upstream roles to {[role.name for role in roles]}"
+        )
+        return roles
+
+    def get_user(self, request: HTTPServerRequest) -> Optional[User]:
+        """Gets the User corresponding to the username and password supplied in the
+        request body
+
+        Args:
+            request: tornado HTTPServerRequest object
+
+        Returns:
+            User: The User object matching the supplied username and password
+            None: If no User was found or the supplied password was invalid
+        """
+        authenticated_user = None
+
+        if request.body:
+            schema = TokenInputSchema()
+
+            request_data = schema.loads(request.body.decode("utf-8")).data
+            username = request_data.get("username")
+            password = request_data.get("password")
+
+            if username and password:
+                try:
+                    server = Server(
+                        host=config.get("ldap.host"),
+                        port=config.get("ldap.port"),
+                        use_ssl=config.get("ldap.use_ssl"),
+                    )
+                    with Connection(
+                        server,
+                        self.get_user_dn(username),
+                        password,
+                    ) as conn:
+                        if self.verify_ldap_password(conn):
+                            try:
+                                authenticated_user = get_user(username=username)
+                            except DoesNotExist:
+                                authenticated_user = User(
+                                    username=username, is_remote=True
+                                )
+                                authenticated_user = create_user(authenticated_user)
+                            authenticated_user.metadata["last_authentication"] = (
+                                datetime.now(timezone.utc).timestamp()
+                            )
+                            authenticated_user.upstream_roles = self.get_user_roles(
```

Comment:
`authenticated_user.roles=self.get_user_roles(`

Since we are using the BG managed roles, we should assign it to the `roles` field. Trusted Headers can get away with providing upstream roles because it could provide JSON blobs for them but LDAP doesn't have the ability to generate a object, unless we do some crazy parsing.

Reply:
complete